### PR TITLE
[FIX] View create permissions

### LINKED ISSUE #2165: View/Create permissions not being applied
- Your Rocket.Chat Experimental app version: 4.7.0.17188
- Your Rocket.Chat server version: 3.3.0
- Deviceyou're running with: Pocophone F1 with Android 10

Steps to reproduce:
- Go to Admin/Permissions and remove the `create-c`, `create-d`, `create-p`, `view-c` and `view-d` permissions for a role such as `livechat-agent`
- Edit a user by giving it **only** the `livechat-agent` role
- Now if you login as this livechat-agent user:
  - In the web application, the icon for creating groups/channels/direct messages doesn't appear, and it can only see private rooms he belongs to (and livechats).
  - In the mobile applications (either in Rocket.Chat or Rocket.Chat Experimental), the permissions are not being applied and the user can still create rooms/channels/direct messages and talk to other users.

One question that comes to mind is, are the permissions not being controlled on the server side, just on the client side?

### PR DESCRIPTION
**Pull request by Jose96GIT**
@RocketChat/ReactNative
Closes #2165

The following screenshots are from an user that has `view-l` and `create-c` permissions.

image-db78ba61-f31b-4132-bdad-5207242e06b7_censored
image-1a15cbd4-4918-4a8d-bca4-4811e9eef716

### CODE REVIEW

**Comment by diegolmello** [comment]
Thanks for contributing, @Jose96GIT!
We'll take a look at it soon.
In the meantime, you can fix the lint :)

**Comment by Jose96GIT** [comment]
I have fixed lint problems, excepting `ios-build` because I don't know how can I fix them.
Otherwise, it is having a merge conflict now, do you want me to fix it now, or should I fix it better after your revision?

**Comment by diegolmello** [comment]
> I have fixed lint problems, excepting `ios-build` because I don't know how can I fix them.
> Otherwise, it is having a merge conflict now, do you want me to fix it now, or should I fix it better after your revision?

@Jose96GIT Please resolve the conflict.
I'm going to review it right after.
Don't worry about the `ios-build`.
Thanks!

**Comment by Jose96GIT** [comment]
The conflict is solved!

File: app/views/RoomsListView/index.js [hunk 1]
```diff
@@ -807,8 +819,10 @@ class RoomsListView extends React.Component {
 		} = this.props;
 		const id = this.getUidDirectMessage(item);
 		const isGroupChat = RocketChat.isGroupChat(item);
-
-		return (
+		const hasViewChannelPermission = item.t !== 'd' ? true : permissions['view-d-room'];
```
[the remark this answers is not in the document]
Comment (diegolmello) [reply app/views/RoomsListView/index.js:822]: Same filter must be added on `unreadsCount` label here: https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/views/RoomView/index.js#L590

It's added on iOS back button like this:



See the `3` unread count on the back button above?
Those 3 unread messages are from a room the user doesn't have access.

[the remark this answers is not in the document]
Comment (Jose96GIT) [reply app/views/RoomsListView/index.js]: I have done what you have suggested here, I cannot be sure that the unread count on back button is right 'cause I don't have an iOS device.

File: app/lib/rocketchat.js [hunk 2]
```diff
@@ -996,6 +996,25 @@ const RocketChat = {
 			log(e);
 		}
 	},
+	async hasPermissionsByUserRoles(permissions = [], userRoles = []) {
```
Comment (djorkaeffalexandre) [inline app/lib/rocketchat.js:999]: To avoid some [duplicated logic](https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/lib/rocketchat.js#L978), try to extract some of this logic to another function that can be reused when we're looking for permissions on user roles or room+user roles.

File: app/views/RoomsListView/index.js [hunk 3]
```diff
@@ -354,14 +364,16 @@ class RoomsListView extends React.Component {
...
 				<CustomHeaderButtons>
-					<Item
-						title='new'
-						iconName='new-chat'
-						onPress={isMasterDetail
-							? () => navigation.navigate('ModalStackNavigator', { screen: 'NewMessageView' })
-							: () => navigation.navigate('NewMessageStackNavigator')}
-						testID='rooms-list-view-create-channel'
-					/>
+					{hasCreateChannelPermission || hasCreateDirectMessagePermission || hasCreatePrivateGroupPermission ? (
```
Comment (djorkaeffalexandre) [inline app/views/RoomsListView/index.js:367]: I guess this button shouldn't be hide when we can't create chat's, since the `NewMessageView` is a "Contacts List" too. WDYT?

File: app/views/RoomView/index.js [hunk 4]
```diff
@@ -427,6 +428,11 @@ class RoomView extends React.Component {
 			const member = await this.getRoomMember();
 
 			this.setState({ canAutoTranslate, member, loading: false });
+
+			const hasPermissions = ['create-c', 'create-d', 'create-p', 'view-c-room', 'view-d-room', 'view-l-room'];
```
Comment (djorkaeffalexandre) [inline app/views/RoomView/index.js:432]: `create-*` permissions aren't used here.

File: app/views/RoomView/index.js [hunk 5]
```diff
@@ -599,7 +605,15 @@ class RoomView extends React.Component {
...
-			const { unreadsCount } = this.state;
+			const { unreadsCount, permissions } = this.state;
+
+			data = data.filter((chat) => {
+				const hasViewChannelPermission = chat.t !== 'd' ? true : permissions['view-d-room'];
+				const hasViewGroupPermission = chat.t !== 'c' ? true : permissions['view-c-room'];
+				const hasViewLivechatPermission = chat.t !== 'l' ? true : permissions['view-l-room'];
+
+				return !(!hasViewGroupPermission || !hasViewChannelPermission || !hasViewLivechatPermission);
+			});
```
Comment (djorkaeffalexandre) [inline app/views/RoomView/index.js:616]: Do this filter logic as a query on database.

File: app/views/NewMessageView.js [hunk 6]
```diff
@@ -179,32 +188,36 @@ class NewMessageView extends React.Component {
 
 	renderHeader = () => {
 		const { maxUsers, theme } = this.props;
-		return (
+		const { permissions } = this.state;
+		const hasCreateChannelPermission = permissions['create-c'];
+		const hasCreateDirectMessagePermission = permissions['create-d'];
+		const hasCreatePrivateGroupPermission = permissions['create-p'];
+		return hasCreateChannelPermission || hasCreateDirectMessagePermission || hasCreatePrivateGroupPermission ? (
```
Comment (djorkaeffalexandre) [inline app/views/NewMessageView.js:195]: this line can be substituted by a class get property like:
```javascript
get showButtons() {
  const { permissions } = this.state;
  ...
  return ...;
}
...
this.showButtons ? () : ()
```

File: app/views/NewMessageView.js [hunk 7]
```diff
@@ -179,32 +188,36 @@ class NewMessageView extends React.Component {
...
 			<View style={{ backgroundColor: themes[theme].auxiliaryBackground }}>
 				<SearchBox onChangeText={text => this.onSearchChangeText(text)} testID='new-message-view-search' />
 				<View style={styles.buttonContainer}>
-					{this.renderButton({
+					{hasCreateChannelPermission ? this.renderButton({
```
Comment (djorkaeffalexandre) [inline app/views/NewMessageView.js:199]: WDYT about create a new `renderButton` property `showButton`:
`showButton: permissions['create-c']`
```javascript
renderButton = ({ showButton }) => {
  if (!showButton) {
    return null;
  }
}
```

Comment (Jose96GIT) [reply app/views/NewMessageView.js:199]: Sounds good!

File: app/views/RoomsListView/index.js (hunk 3 above)
Comment (Jose96GIT) [reply app/views/RoomsListView/index.js:367]: You're right.

To achieve a real Contact List, I've disabled the possibility to enter to the chat when the user has no view permission.

**Comment by Jose96GIT** [comment]
@djorkaeffalexandre thanks for your code revision, I've solved all threads.

I have also added the filter to the `search bar` and `DirectoryView` but I didn't notice any difference in the `DirectoryView` because it was displaying me any direct room.

File: app/views/RoomsListView/index.js [hunk 8]
```diff
@@ -189,6 +190,10 @@ class RoomsListView extends React.Component {
 				this.backHandler.remove();
 			}
 		});
+		const hasPermissions = ['create-c', 'create-d', 'create-p', 'view-c-room', 'view-d-room', 'view-l-room'];
+		const permissions = await RocketChat.hasPermissionsByUserRoles(hasPermissions);
+		this.setState({ permissions });
+		this.getSubscriptions(true);
```
Comment (djorkaeffalexandre) [inline app/views/RoomsListView/index.js:196]: just add a new condition on `componentDidUpdate` to check if `permissions` changed on [this conditional](https://github.com/RocketChat/Rocket.Chat.ReactNative/blob/develop/app/views/RoomsListView/index.js#L285) instead of call `this.getSubscriptions` here.

Comment (djorkaeffalexandre) [inline app/views/RoomsListView/index.js:196]: Extract this to other function like `setPermissions` and call here.

File: app/views/RoomView/index.js [hunk 9]
```diff
@@ -588,18 +592,26 @@ class RoomView extends React.Component {
...
 				Q.where('open', true),
-				Q.where('rid', Q.notEq(this.rid))
+				Q.where('rid', Q.notEq(this.rid)),
+				Q.where('t', Q.notIn(notPermissionsChat))
 			)
 			.observeWithColumns(['unread']);
 
 		this.queryUnreads = observable.subscribe((data) => {
 			const { unreadsCount } = this.state;
+
```
Comment (djorkaeffalexandre) [inline app/views/RoomView/index.js:614]: Unnecessary change

File: app/views/RoomsListView/index.js [hunk 10]
```diff
@@ -189,6 +190,10 @@ class RoomsListView extends React.Component {
 				this.backHandler.remove();
 			}
 		});
+		const hasPermissions = ['create-c', 'create-d', 'create-p', 'view-c-room', 'view-d-room', 'view-l-room'];
```
Comment (djorkaeffalexandre) [inline app/views/RoomsListView/index.js:193]: `create-*` aren't being used

File: app/views/NewMessageView.js [hunk 11]
```diff
@@ -213,9 +239,20 @@ class NewMessageView extends React.Component {
 	}
 
 	renderItem = ({ item, index }) => {
-		const { search, chats } = this.state;
+		const { search, chats, permissions } = this.state;
 		const { baseUrl, user, theme } = this.props;
 
+		let hasPermissions = false;
```
Comment (djorkaeffalexandre) [inline app/views/NewMessageView.js:245]: `NewMessagesView` is used essential to display only Direct Message rooms, we don't need to check all these room types.

File: app/views/RoomsListView/index.js [hunk 12]
```diff
@@ -415,12 +420,19 @@ class RoomsListView extends React.Component {
 			groupByType
 		} = this.props;
 
+		const { permissions } = this.state;
+		const notPermissionsChat = [];
+		if (!permissions['view-d-room']) { notPermissionsChat.push('d'); }
+		if (!permissions['view-c-room']) { notPermissionsChat.push('c'); }
+		if (!permissions['view-l-room']) { notPermissionsChat.push('l'); }
```
Comment (djorkaeffalexandre) [inline app/views/RoomsListView/index.js:427]: Try to do this logic inside some function that will run only once.

File: app/views/DirectoryView/index.js [hunk 13]
```diff
@@ -181,9 +185,16 @@ class DirectoryView extends React.Component {
 	}
 
 	renderItem = ({ item, index }) => {
-		const { data, type } = this.state;
+		const { data, type, permissions } = this.state;
```
[the remark this answers is not in the document]
Comment (Jose96GIT) [reply app/views/DirectoryView/index.js]: You're right, this is a better way to do it.

**Comment by Jose96GIT** [comment]
@djorkaeffalexandre I've just finished with your last changes. Thanks again for your review!

**Comment by Jose96GIT** [comment]
Good morning! Is there any other change that should I do before merging this fix? Sorry for the inconvenience.

**Comment by diegolmello** [comment]
> Good morning! Is there any other change that should I do before merging this fix? Sorry for the inconvenience.

@Jose96GIT Np. We're releasing a new version, so we don't want to merge it right now.
We're probably going to do it next week.
Don't worry about keeping it updated.
We'll resolve the conflicts, if necessary.
Thanks.

**Comment by yash-rajpal** [comment]
@diegolmello Should I make a new PR for this and removing all conflicts?

**Comment by diegolmello** [comment]
Some of this was implemented on #3233
